File: FlexServ Backend/SpringBoot/AuthService/src/main/java/com/flexserv/service/AuthServiceImpl.java
package com.flexserv.service;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.flexserv.dto.auth.AuthenticationResult;
import com.flexserv.dto.request.LoginRequest;
import com.flexserv.dto.request.RegisterRequest;
import com.flexserv.dto.response.LoginResponse;
import com.flexserv.dto.response.UserResponse;
import com.flexserv.entity.Role;
import com.flexserv.entity.ServiceProviderCompany;
import com.flexserv.entity.User;
import com.flexserv.exception.InvalidCredentialsException;
import com.flexserv.exception.UserAlreadyExistsException;
import com.flexserv.repository.ServiceProviderCompanyRepository;
import com.flexserv.repository.UserRepository;
import com.flexserv.security.JwtService;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AuthServiceImpl implements AuthService {

	private final UserRepository userRepository;
	private final ServiceProviderCompanyRepository serviceProviderCompanyRepository;
	private final PasswordEncoder passwordEncoder;
	private final JwtService jwtService;

	@Override
	@Transactional
	public UserResponse register(RegisterRequest request) {

		if (userRepository.existsByEmail(request.getEmail())) {
			throw new UserAlreadyExistsException("Email already exists");
		}

		if (userRepository.existsByPhone(request.getPhone())) {
			throw new UserAlreadyExistsException("Phone number already exists");
		}

		User user = new User();

		user.setName(request.getName());
		user.setEmail(request.getEmail());
		user.setPhone(request.getPhone());
		user.setPassword(passwordEncoder.encode(request.getPassword()));

		Role role = request.getRole();
		if (role == null) {
			role = Role.CUSTOMER;
		}
		user.setRole(role);

		User savedUser = userRepository.save(user);

		// If registering as Service Provider, create associated ServiceProviderCompany
		if (role == Role.PROVIDER || role == Role.provider || "PROVIDER".equalsIgnoreCase(role.name())) {
			ServiceProviderCompany company = new ServiceProviderCompany();
			company.setUser(savedUser);
			company.setCompanyName(
				request.getCompanyName() != null && !request.getCompanyName().isBlank()
					? request.getCompanyName()
					: savedUser.getName() + " Services"
			);
			company.setExperienceYears(
				request.getExperienceYears() != null ? request.getExperienceYears() : 1
			);
			company.setBio(
				request.getBio() != null && !request.getBio().isBlank()
					? request.getBio()
					: "Professional service provider registered on FlexServ."
			);
			company.setIsVerified(true);
			company.setRating(5.0);
			company.setCompanyAvailable(true);

			serviceProviderCompanyRepository.save(company);
		}

		UserResponse response = new UserResponse();
		response.setId(savedUser.getId());
		response.setName(savedUser.getName());
		response.setEmail(savedUser.getEmail());
		response.setPhone(savedUser.getPhone());
		response.setRole(savedUser.getRole());
		response.setCreatedAt(savedUser.getCreatedAt());

		return response;
	}

	@Override
	public AuthenticationResult login(LoginRequest request) {

		User user = userRepository.findByEmail(request.getEmail())
				.orElseThrow(() ->
						new InvalidCredentialsException("Invalid email or password"));

		if (Boolean.FALSE.equals(user.getIsActive())) {
			throw new com.flexserv.exception.AccountDeactivatedException(
					"Your account has been deactivated / soft-deleted by administration. Access denied."
			);
		}

		if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
			throw new InvalidCredentialsException("Invalid email or password");
		}

		String token = jwtService.generateToken(user);

		LoginResponse loginResponse = new LoginResponse(
				user.getId(),
				user.getName(),
				user.getRole(),
				token
		);

		return new AuthenticationResult(token, loginResponse);
	}
}

File: FlexServ Backend/DotNet/AuthService/Services/AuthServiceImpl.cs
using AuthService.Data;
using AuthService.Dtos;
using AuthService.Models;
using System;
using System.Linq;

namespace AuthService.Services
{
    public class AuthServiceImpl : IAuthService
    {
        private readonly AuthDbContext _dbContext;
        private readonly IJwtService _jwtService;

        public AuthServiceImpl(AuthDbContext dbContext, IJwtService jwtService)
        {
            _dbContext = dbContext;
            _jwtService = jwtService;
        }

        public UserResponse Register(RegisterRequest request)
        {
            if (_dbContext.Users.Any(u => u.Email == request.Email))
            {
                throw new InvalidOperationException("Email already exists");
            }

            if (_dbContext.Users.Any(u => u.Phone == request.Phone))
            {
                throw new InvalidOperationException("Phone number already exists");
            }

            var user = new User
            {
                Name = request.Name,
                Email = request.Email,
                Phone = request.Phone,
                Password = BCrypt.Net.BCrypt.HashPassword(request.Password),
                Role = request.Role ?? Role.CUSTOMER,
                IsActive = true,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            _dbContext.Users.Add(user);
            _dbContext.SaveChanges();

            return new UserResponse
            {
                Id = user.Id,
                Name = user.Name,
                Email = user.Email,
                Phone = user.Phone,
                Role = user.Role,
                Active = user.IsActive,
                CreatedAt = user.CreatedAt
            };
        }

        public AuthenticationResult Login(LoginRequest request)
        {
            var user = _dbContext.Users.FirstOrDefault(u => u.Email == request.Email);
            if (user == null || !BCrypt.Net.BCrypt.Verify(request.Password, user.Password))
            {
                throw new InvalidOperationException("Invalid email or password");
            }

            var token = _jwtService.GenerateToken(user);
            var loginResponse = new LoginResponse(user.Id, user.Name, user.Role);

            return new AuthenticationResult(token, loginResponse);
        }
    }
}
